docs/trpc-examples.md

```diff
@@ -0,0 +1,831 @@
+
+  firstName String
+  lastName  String
+  user      User?
+}
+
+model User {
+  id        String   @id @default(cuid())
+  createdAt DateTime @default(now())
+  updatedAt DateTime @updatedAt
+  email     String   @unique
+
+  profileId String
+  profile   Profile  @relation(fields: [profileId], references: [id])
+  skills    Skill[]
+  posts     Post[]
+}
+
+model Skill {
+  id          String   @id @default(cuid())
+  createdAt   DateTime @default(now())
+  updatedAt   DateTime @updatedAt
+
+  name        String   @unique
+  description String?
+  archived    Boolean  @default(false)
+  users       User[]
+}
+
+model Post {
+  id        String   @id @default(cuid())
+  createdAt DateTime @default(now())
+  updatedAt DateTime @updatedAt
+
+  title     String   @unique
+  body      String?
+  userId    String
+  user      User     @relation(fields: [userId], references: [id])
+}
+
+```
+
+</details>
+
+## Router References
+
+<details>
+<summary>
+  Full User Module Example
+</summary>
+
+```typescript
+import { Prisma } from '@prisma/client';
+import { z } from 'zod';
+
+import { t } from '@/server/trpc';
+import { adminProcedure, protectedProcedure } from '@/server/middleware/auth';
+
+import { isAdmin } from '../../services/permissions';
+import { appJwtForUser, comparePasswords, hashPassword } from '../../services/auth';
+
+export const defaultUserSelect = Prisma.validator<Prisma.UserSelect>()({
+  id: true,
+  email: true,
+});
+
+// User Type
+export const User = objectType({
+  name: 'User',
+  description: 'A User',
+  definition(t) {
+    t.nonNull.id('id');
+    t.nonNull.date('createdAt');
+    t.nonNull.date('updatedAt');
+    t.nonNull.string('email');
+
+    t.nonNull.list.nonNull.field('skills', {
+      type: 'Skill',
+      resolve: async (parent, _, context) => {
+        return context.prisma.user
+          .findUnique({
+            where: { id: parent.id },
+          })
+          .skills();
+      },
+    });
+
+    t.field('profile', {
+      type: 'Profile',
+      resolve: (parent, _, context) => {
+        return context.prisma.user
+          .findUnique({
+            where: { id: parent.id },
+          })
+          .profile();
+      },
+    });
+
+    t.list.field('posts', {
+      type: 'Post',
+      resolve: async (parent, _, context) => {
+        return context.prisma.user
+          .findUnique({
+            where: { id: parent.id },
+          })
+          .posts();
+      },
+    });
+  },
+});
+
+// Auth Payload Type
+export const AuthPayload = objectType({
+  name: 'AuthPayload',
+  description: 'Payload returned if login is successful',
+  definition(t) {
+    t.field('user', { type: 'User', description: 'The logged in user' });
+    t.string('token', {
+      description: 'The current JWT token. Use in Authentication header',
+    });
+  },
+});
+
+export const userRouter = t.router({
+  me: protectedProcedure.query(async ({ ctx }) => ctx.user),
+  users: protectedProcedure
```

Comment:
I know this is just documentation, but as a general convention, do you think it's worth putting a comment above each procedure explaining what it is?

Example:

```ts
/*
   Returns a list of users with profiles. Ordered name desc by default.
*/
```

Most of these are straightforward by reading through or with really descriptive procedure names, but it could help with scanning the file and when autocompleting with TS?

Reply:
Yeah, or we could name the procedures better. The entire router is called `user` so the procedures could be called `getAll`, `getById` or `getMe`, and accessed like

```
trpc.user.getAll.useQuery()
trpc.user.getById.useQuery()
trpc.user.getMe.useQuery()
```

We've got options, I'm happy to adjust or add comments.

Reply:
We can also do this is a followup. just came to mind as I was reviewing. I feel like the right answer is probably both in some capacity.